firedrake/preconditioners/pmg.py

```diff
@@ -0,0 +1,300 @@
+    will subclass :class:`PMGPC` to override `coarsen_element`.
+    """
+    @staticmethod
+    def coarsen_element(ele):
+        """
+        Coarsen a given element to form the next problem down in the p-hierarchy.
+
+        If the supplied element should form the coarsest level of the p-hierarchy,
+        raise `ValueError`. Otherwise, return a new :class:`ufl.FiniteElement`.
+
+        By default, this does power-of-2 coarsening in polynomial degree.
+        It raises a `NotImplementedError` for :class:`ufl.MixedElement`s, as
+        I don't know if there's a sensible default strategy to implement here.
+        It is intended that the user subclass `PMGPC` to override this method
+        for their problem.
+
+        :arg ele: a :class:`ufl.FiniteElement` to coarsen.
+        """
+        if isinstance(ele, MixedElement) and not isinstance(ele, (VectorElement, TensorElement)):
+            raise NotImplementedError("Implement this method yourself")
+
+        degree = ele.degree()
+        family = ele.family()
+
+        if family == "Discontinuous Galerkin" and degree == 0:
+            raise ValueError
+        elif degree == 1:
+            raise ValueError
+
+        shape = ele.value_shape()
+
+        if len(shape) == 0:
+            new_ele = FiniteElement(family, ele.cell(), degree // 2)
+        elif len(shape) == 1:
+            new_ele = VectorElement(family, ele.cell(), degree // 2, dim=shape[0])
+        else:
+            new_ele = TensorElement(family, ele.cell(), degree // 2, shape=shape,
+                                    symmetry=ele.symmetry())
+        return new_ele
+
+    def initialize(self, pc):
+        # Make a new DM.
+        # Hook up a (new) coarsen routine on that DM.
+        # Make a new PC, of type MG.
+        # Assign the DM to that PC.
+
+        odm = pc.getDM()
+        ctx = get_appctx(odm)
+
+        test, trial = ctx.J.arguments()
+        if test.function_space() != trial.function_space():
+            raise NotImplementedError("test and trial spaces must be the same")
+
+        # Construct a list with the elements we'll be using
+        V = test.function_space()
+        ele = V.ufl_element()
+        elements = [ele]
+        while True:
+            try:
+                ele = self.coarsen_element(ele)
+            except ValueError:
+                break
+            elements.append(ele)
+
+        pdm = PETSc.DMShell().create(comm=pc.comm)
+        sf = odm.getPointSF()
+        section = odm.getDefaultSection()
+        attach_hooks(pdm, level=len(elements)-1, sf=sf, section=section)
+        # Now overwrite some routines on the DM
+        pdm.setRefine(None)
+        pdm.setCoarsen(self.coarsen)
+        pdm.setCreateInterpolation(self.create_interpolation)
+        set_function_space(pdm, get_function_space(odm))
+
+        parent = get_parent(odm)
+        add_hook(parent, setup=partial(push_parent, pdm, parent), teardown=partial(pop_parent, pdm, parent),
```

Comment:
Should we `assert parent is not None`?

Reply:
Don't recall if that is necessary.

Reply:
I added the assertion, I don't think it does any harm.